xbmc/FileItem.cpp

```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
-        j = i + 1;
-        while (j < Size())
-        {
-          CFileItemPtr item2 = Get(j);
-
-          // skip folders, nfo files, playlists
-          if (item2->m_bIsFolder
-            || item2->IsParentFolder()
-            || item2->IsNFO()
-            || item2->IsPlayList()
-            )
-          {
-            // increment index
-            j++;
-            continue;
-          }
+        // Find all items in stack
+        std::vector<int> stack;
+        int64_t size{0};
 
-          std::string file2, filePath2;
-          URIUtils::Split(item2->GetPath(), filePath2, file2);
-          if (URIUtils::HasEncodedFilename(CURL(filePath2)) )
-            file2 = CURL::Decode(file2);
-
-          if (expr->RegFind(file2, offset) != -1)
+        for (const auto& stackItem : stackCandidates)
+        {
+          if (stackBaseItem.first == std::get<0>(stackItem) && // type
+              stackBaseItem.second == std::get<1>(stackItem)) // title
           {
-            std::string  Title2      = expr->GetMatch(1),
-                        Volume2     = expr->GetMatch(2),
-                        Ignore2     = expr->GetMatch(3),
-                        Extension2  = expr->GetMatch(4);
-            if (offset)
-              Title2 = file2.substr(0, expr->GetSubStart(2));
-            if (StringUtils::EqualsNoCase(Title1, Title2))
+            const int index{std::get<3>(stackItem)}; // index
+            stack.emplace_back(index);
+            size += std::get<5>(stackItem); // size
+            const std::string playPath = std::get<4>(stackItem); // updated path
+            if (!playPath.empty())
             {
-              if (!StringUtils::EqualsNoCase(Volume1, Volume2))
-              {
-                if (StringUtils::EqualsNoCase(Ignore1, Ignore2) &&
-                    StringUtils::EqualsNoCase(Extension1, Extension2))
-                {
-                  if (stack.empty())
-                  {
-                    stackName = Title1 + Ignore1 + Extension1;
-                    stack.push_back(i);
-                    size += item1->m_dwSize;
-                  }
-                  stack.push_back(j);
-                  size += item2->m_dwSize;
-                }
-                else // Sequel
-                {
-                  offset = 0;
-                  ++expr;
-                  break;
-                }
-              }
-              else if (!StringUtils::EqualsNoCase(Ignore1, Ignore2)) // False positive, try again with offset
-              {
-                offset = expr->GetSubStart(3);
-                break;
-              }
-              else // Extension mismatch
-              {
-                offset = 0;
-                ++expr;
-                break;
-              }
+              CFileItemPtr item{Get(index)};
+              item->SetPath(playPath); // updated path (for DVD/Bluray files)
+              item->m_bIsFolder = false;
+              item->SetLabel2("");
```

Comment:
This could be rewritten to use a bit more modern C++, and (more important) we want to avoid setting item's path after construction, because this semantically is a change of identity, which strictly speaking must not happen.


```c++
const auto item{std::make_shared<CFileItem>(playPath, false)};
item->SetLabel(Get(index));
```

No need to set an empty label2, btw, as this is the default.

Reply:
Done

Reply:
Needs to be `const std::shared_ptr<CFileItem> item{Get(index)};`